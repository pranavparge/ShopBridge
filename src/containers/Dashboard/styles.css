.table {
  margin-top: 1em;
  margin-bottom: 1em;
}

@media only screen and (min-width: 600px) {
  .addBtn {
    float: right;
    margin-bottom: 1em;
  }
  .addBtnMobile {
    display: none;
  }
}

@media only screen and (max-width: 600px) {
  .addBtn {
    display: none;
  }
  .addBtnMobile {
    width: 100%;
  }
}

.spinner {
  width: 4em;
  height: 4em;
}

.loader {
  margin-left: 50%;
  margin-top: 10%;
}

.btnGrp {
  width: 10%;
}
