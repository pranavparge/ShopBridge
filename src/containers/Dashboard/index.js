import React, { useEffect, useState } from "react";
import { Button, Table, Alert, Spinner, ButtonGroup } from "react-bootstrap";

import Modal from "../../components/Modal";

import "./styles.css";

function Dashboard() {
  const [itemsResponse, setItemsResponse] = useState(null);
  const [itemsErrors, setItemsErrors] = useState(false);

  const [itemDeleteErrors, setItemDeleteErrors] = useState(false);
  const [itemUpdateErrors, setItemupdateErrors] = useState(false);

  const [itemAddResponse, setItemAddResponse] = useState(false);
  const [itemAddErrors, setItemAddErrors] = useState(false);

  const [modalType, setModalType] = useState(null);
  const [itemToUpdate, setItemToUpdate] = useState(null);

  const [show, setShow] = useState(false);
  const [showAlert, setShowAlert] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = (action, item) => {
    setShow(true);
    setModalType(action);
    setItemToUpdate(item);
  };

  const fetchItems = () => {
    fetch("http://localhost:3001/items")
      .then((response) => response.json())
      .then((response) => {
        setItemsResponse(response);
        setItemsErrors(false);
      })
      .catch(() => {
        setItemsResponse(null);
        setItemsErrors(true);
      });
  };

  const handleAdd = (item) => {
    fetch("http://localhost:3001/items", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({
        id: item.id,
        name: item.name,
        description: item.description,
        price: item.price,
        quantity: item.quantity,
      }),
    })
      .then((response) => response.json())
      .then((response) => {
        setItemAddResponse(response);
        setItemAddErrors(false);
      })
      .then(() => fetchItems())
      .catch((err) => {
        setItemAddResponse("Duplicate ID");
        setItemAddErrors(true);
      });
  };

  const handleDelete = (item) => {
    fetch("http://localhost:3001/items/" + item.id, { method: "DELETE" })
      .then(() => {
        setItemDeleteErrors(false);
      })
      .then(() => fetchItems())
      .catch(() => {
        setItemDeleteErrors(true);
      });
  };

  const handleUpdate = (updatedItem) => {
    fetch("http://localhost:3001/items/" + updatedItem.id, {
      method: "PUT",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({
        id: updatedItem.id,
        name: updatedItem.name,
        description: updatedItem.description,
        price: updatedItem.price,
        quantity: updatedItem.quantity,
      }),
    })
      .then(() => {
        setItemupdateErrors(false);
      })
      .then(() => fetchItems())
      .catch(() => {
        setItemupdateErrors(true);
      });
  };

  useEffect(() => {
    fetchItems();
  }, []);

  useEffect(() => {
    if (
      (itemsErrors && itemsResponse === null) ||
      (itemAddErrors && itemAddResponse) ||
      itemDeleteErrors ||
      itemUpdateErrors
    ) {
      setShowAlert(true);
    } else {
      setShowAlert(false);
    }
  }, [
    itemsErrors,
    itemsResponse,
    itemAddErrors,
    itemAddResponse,
    itemDeleteErrors,
    itemUpdateErrors,
  ]);

  return (
    <div className="Dashboard">
      <h1>Inventory</h1>

      <Modal
        {...{
          show,
          handleClose,
          modalType,
          itemToUpdate,
          handleAdd,
          handleUpdate,
        }}
      ></Modal>

      {showAlert && (
        <Alert onClose={() => setShowAlert(false)} dismissible variant="danger">
          {itemAddErrors
            ? itemAddResponse.toString()
            : "An Error Occurred! Sorry!"}
        </Alert>
      )}

      {(itemsResponse && !itemsErrors && (
        <>
          <Button
            className="addBtn"
            variant="primary"
            onClick={() => handleShow("add", null)}
          >
            Add Item
          </Button>
          <Table className="table" bordered hover>
            <thead>
              <tr>
                <th>Action</th>
                <th>#</th>
                <th>Name</th>
                <th>Description</th>
                <th>Price</th>
                <th>Quantity</th>
              </tr>
            </thead>
            <tbody>
              {itemsResponse.map((item) => (
                <tr key={item.id}>
                  <td className="btnGrp">
                    <ButtonGroup>
                      <Button
                        variant="danger"
                        onClick={() => handleDelete(item)}
                      >
                        Delete
                      </Button>
                      <Button
                        variant="warning"
                        onClick={() => handleShow("update", item)}
                      >
                        Update
                      </Button>
                    </ButtonGroup>
                  </td>
                  <td>{item.id}</td>
                  <td>{item.name}</td>
                  <td>{item.description}</td>
                  <td>{item.price && item.price + "â‚¹"}</td>
                  <td>{item.quantity}</td>
                </tr>
              ))}
            </tbody>
          </Table>

          <Button className="addBtnMobile" variant="primary">
            Add Item
          </Button>
        </>
      )) || (
        <div className="loader">
          <Spinner className="spinner" animation="border" variant="primary" />
        </div>
      )}
    </div>
  );
}

export default Dashboard;
