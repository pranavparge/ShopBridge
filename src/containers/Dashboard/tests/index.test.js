import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import Dashboard from "../index";

const component = <Dashboard />;

test("to check if it matches snapshot", () => {
  const { asFragment } = render(component);
  expect(asFragment(component)).toMatchSnapshot();
});

test("to check if items get loaded", async () => {
  render(component);
  await waitFor(() =>
    expect(screen.getByTestId("itemsTable")).toBeInTheDocument()
  );
});

test("to check if modal closes", async () => {
  render(component);

  fireEvent.click(await screen.findByTestId("addItemBtn"));
  const modal = await screen.findByTestId("modal");
  await waitFor(() => {
    expect(modal).toBeInTheDocument();
    expect(screen.getByTestId("cancelBtn")).toBeInTheDocument();
    fireEvent.click(screen.getByTestId("cancelBtn"));
  });

  await waitFor(() => {
    expect(modal).not.toBeInTheDocument();
  });
});

test("to check if item can be added", async () => {
  render(component);
  fireEvent.click(await screen.findByTestId("addItemBtn"));

  await waitFor(() => {
    expect(screen.getByTestId("modal")).toBeInTheDocument();

    const itemIDInput = screen.getByTestId("itemID");
    fireEvent.change(itemIDInput, { target: { value: "100" } });
    expect(itemIDInput.value).toBe("100");

    const itemNameInput = screen.getByTestId("itemName");
    fireEvent.change(itemNameInput, { target: { value: "Name" } });
    expect(itemNameInput.value).toBe("Name");

    const itemDescInput = screen.getByTestId("itemDesc");
    fireEvent.change(itemDescInput, { target: { value: "Desc" } });
    expect(itemDescInput.value).toBe("Desc");

    const itemPriceInput = screen.getByTestId("itemPrice");
    fireEvent.change(itemPriceInput, { target: { value: "12" } });
    expect(itemPriceInput.value).toBe("12");

    const itemQuantityInput = screen.getByTestId("itemQuantity");
    fireEvent.change(itemQuantityInput, { target: { value: "100" } });
    expect(itemQuantityInput.value).toBe("100");

    fireEvent.click(screen.getByTestId("applyBtn"));
  });
});

test("to check if item can be updated", async () => {
  render(component);

  const element = await screen.findByTestId("updateBtn_100");
  expect(element).toBeInTheDocument();
  fireEvent.click(element);

  await waitFor(() => {
    expect(screen.getByTestId("modal")).toBeInTheDocument();

    const itemQuantityInput = screen.getByTestId("itemQuantity");
    fireEvent.change(itemQuantityInput, { target: { value: "200" } });
    expect(itemQuantityInput.value).toBe("200");

    fireEvent.click(screen.getByTestId("applyBtn"));
  });

  await waitFor(() => {
    const updatedElement = screen.getByTestId("quantity_100");
    expect(updatedElement.innerHTML).toBe("200");
  });
});

test("to check if item can be deleted", async () => {
  render(component);

  const element = await screen.findByTestId("deleteBtn_100");
  expect(element).toBeInTheDocument();
  fireEvent.click(element);
  await waitFor(() => {
    expect(element).not.toBeInTheDocument();
  });
});
