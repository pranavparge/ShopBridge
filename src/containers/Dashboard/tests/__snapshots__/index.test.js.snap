// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to check if it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Dashboard"
  >
    <h1>
      Inventory
    </h1>
    <div
      class="loader"
    >
      <div
        class="spinner spinner-border text-primary"
      />
    </div>
  </div>
</DocumentFragment>
`;
